pkg/systemd/reporting.jsx

```diff
@@ -0,0 +1,463 @@
+ */
+
+import cockpit from "cockpit";
+import React from "react";
+import { show_modal_dialog } from "cockpit-components-dialog.jsx";
+
+const _ = cockpit.gettext;
+
+const TaskState = Object.freeze({
+    READY: 0,
+    RUNNING: 1,
+    COMPLETED: 2,
+    ERROR: 3,
+    CANCELED: 4,
+});
+
+const PromptType = Object.freeze({
+    ASK: 0,
+    ASK_YES_NO: 1,
+    ASK_YES_NO_YESFOREVER: 2,
+    ASK_YES_NO_SAVE: 3,
+    ASK_PASSWORD: 4,
+});
+
+const ProblemState = Object.freeze({
+    REPORTABLE: 0,
+    REPORTING: 1,
+    REPORTED: 2,
+    UNREPORTABLE: 3,
+});
+
+const client = cockpit.dbus("org.freedesktop.problems", { superuser: "try" });
+
+// For one-off fetches of properties to avoid setting up a cache for everything.
+function get_problem_property(problem, property) {
+    const promise = cockpit.defer();
+
+    client.wait().done(() => {
+        client
+                .call(problem.path,
+                      "org.freedesktop.DBus.Properties",
+                      "Get", ["org.freedesktop.Problems2.Entry", property])
+                .done((property) => promise.resolve(property))
+                .fail((ex) => console.error(cockpit.format("Getting property $0 for problem $1 failed: $2", property, problem, ex)));
+    });
+
+    return promise;
+}
+
+class FAFWorkflowRow extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            problemState: ProblemState.REPORTABLE,
+            process: null,
+            reportLinks: [],
+            message: "",
+        };
+
+        this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
+        this.onReportButtonClick = this.onReportButtonClick.bind(this);
+        this.updateStatusFromBus = this.updateStatusFromBus.bind(this);
+
+        this.updateStatusFromBus();
+    }
+
+    updateStatusFromBus() {
+        get_problem_property(this.props.problem, "Reports")
+                .promise()
+                .then((reports) => {
+                    const reportLinks = [];
+                    let reported = false;
+
+                    for (const report of reports[0].v) {
+                        if (report[0] === "ABRT Server") {
+                            if ("URL" in report[1]) {
+                                reportLinks.push(report[1].URL.v.v);
+                            }
+                            reported = true;
+                        }
+                    }
+
+                    this.setState({
+                        problemState: reported ? ProblemState.REPORTED : this.state.problemState,
+                        reportLinks: reportLinks,
+                    });
+                });
+        get_problem_property(this.props.problem, "CanBeReported")
+                .promise()
+                .then((reportable) => {
+                    if (!reportable) {
+                        this.setState({ problemState: ProblemState.UNREPORTABLE });
```

Comment:
This and line 102 look like a race condition. Both properties are being queried in parallel (with two separate D-Bus calls), so depending on which one wins, problemState may end up as UNREPORTABLE or REPORTED. It may be better to do a `.GetAll()` call and look at both properties in the same function, to avoid races and ambiguity?

Reply:
That is actually very unlikely to happen, but maybe fetching everything at once would be better, indeed.